subsys/net/lib/dhcpv4/dhcpv4_server.c

```diff
@@ -752,9 +938,26 @@ static void dhcpv4_handle_discover(struct dhcpv4_server_ctx *ctx,
 	} else {
 		uint32_t lease_time = dhcpv4_get_lease_time(options, optlen);
 
-		if (dhcpv4_send_offer(ctx, msg, &selected->addr, lease_time,
-				      &params) < 0) {
-			return;
+		if (IS_ENABLED(DHCPV4_SERVER_ICMP_PROBE) && probe) {
+			if (dhcpv4_server_probe_setup(ctx, selected,
+						      msg, &params) < 0) {
+				/* Probing context already in use or failed to
+				 * send probe, ignore Discovery for now and wait
+				 * for retransmission.
+				 */
+				return;
+			}
+
+			selected->expiry =
+				sys_timepoint_calc(ADDRESS_PROBE_TIMEOUT);
+		} else {
+			if (dhcpv4_send_offer(ctx, msg, &selected->addr,
+					      lease_time, &params) < 0) {
+				return;
```

Comment:
are we not clearing the flag for that address here like above?

Also as a side note: Sending DHCP offer doesn't mean that client will use that IP address, a client can get multiple offers and can choose one, servers other than the one client has chosen can release:

> The server SHOULD mark an
     address offered to a client in a DHCPOFFER message as available if
     the server receives no DHCPREQUEST message from that client.

Reply:
> are we not clearing the flag for that address here like above?

Which flag are you referring to? If we fail to send offer, the address remains marked as FREE.

> Also as a side note: Sending DHCP offer doesn't mean that client will use that IP address, a client can get multiple offers and can choose one, servers other than the one client has chosen can release:

True, that why the RESERVED state is not permanent - if we don't get a Request for that address, it will time out after 5 seconds and get back to FREE state.

Reply:
> > are we not clearing the flag for that address here like above?
> 
> Which flag are you referring to? If we fail to send offer, the address remains marked as FREE.
Line 827, if sending the offer fails we are doing: `slot->state = DHCPV4_SERVER_ADDR_FREE;`

> > Also as a side note: Sending DHCP offer doesn't mean that client will use that IP address, a client can get multiple offers and can choose one, servers other than the one client has chosen can release:
> 
> True, that why the RESERVED state is not permanent - if we don't get a Request for that address, it will time out after 5 seconds and get back to FREE state.

Yes, missed Line 1050, thanks.